### DISCUSSION TITLE
Netbox randomly doesn't commit changes unless "Commit changes" box is reticked

### DISCUSSION THREAD

**Discussion by Cennepal** [netbox-community/netbox#17037] [Help Wanted!]
Hello, I'm having a very annoying issue with one of my scripts currently. Source code [here](https://github.com/Cennepal/Netbox-IP-Scanner-2/blob/main/netbox_ipscanner_2.py).  

The title already says it and I'm kinda lost on what to do. The script works absolutely fine, the code throws no errors at me. However, it just randomly throws me an "Database changes have been reverted automatically" and doesn't commit the data. Or, with larger scans (~30 subnets), there is no revert message at all, the log looks like the script did its function, but the data still isn't commited.

The only reliable 'fix' for this that I found, is manually reticking the "Commit changes" box before execution, but that's just not optimal for my usecase. All of my other scripts work fine, this one is the only one with this issue.

Any thoughts?

**Comment by candlerb** [comment]
Netbox version?

I wonder if your script is raising an uncaught exception sometimes. But then (a) I'd have thought that would be logged, and (b) unticking and ticking the box wouldn't make any difference.

I can't see any mechanism by which (b) would make any difference, but if you use your browser's console to capture the POST data in both cases (i.e. once without touching the box, and unticking/ticking the box), and you can show a difference between the two, then that could be investigated further.

One thing that looks potentially non-determinstic in your script is `reverse_dns_lookup()`. If you can change this to return a static value, or a value derived directly from the ip address without calling `gethostbyaddr`, does that make a difference?  I'm thinking that maybe some DNS queries are failing, e.g. when TTL expires.

**Reply by Cennepal** [reply]
Also, I meant the logging of the script, as in self.log_info. Does netbox even have deeper logging?

**Reply by Cennepal** [reply]
I'm running 4.0.8. I need to check with the static values, but it will take time since the issue either occurs literally every scan or very rarely. Anyways, I'll update on that when I have results. For the POST requests, here they are:  
```
Failed
-------
Request:
POST /extras/scripts/77/ HTTP/1.1
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.9,de;q=0.8,de-DE;q=0.7,en-GB;q=0.6
Cache-Control: max-age=0
Connection: keep-alive
Content-Length: 90
Content-Type: application/x-www-form-urlencoded
Cookie: django_language=en; csrftoken=RCTDoTVRLlWU75s902mOYSXOdhSSw7PB; sessionid=rypo0vq8c0w292x481l9ngixeb3rbkzm
DNT: 1
Host: docker.domain.ad:8000
Origin: http://docker.domain.ad:8000
Referer: http://docker.domain.ad:8000/extras/scripts/
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36 Edg/127.0.0.0

Response:
HTTP/1.1 302 Found
Content-Type: text/html; charset=utf-8
Location: /extras/scripts/results/160/
X-Request-ID: 99483dac-5514-4ef7-a647-b5f96178f7b7
Vary: HX-Request, Accept-Language, Cookie, origin
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin
Cross-Origin-Opener-Policy: same-origin
X-Frame-Options: SAMEORIGIN
Content-Length: 0
Content-Language: en
access-control-allow-origin: *
Server: Unit/1.32.0
Date: Wed, 31 Jul 2024 09:04:04 GMT

----------
Successful
----------
Request:
POST /extras/scripts/77/ HTTP/1.1
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.9,de;q=0.8,de-DE;q=0.7,en-GB;q=0.6
Cache-Control: max-age=0
Connection: keep-alive
Content-Length: 690
Content-Type: multipart/form-data; boundary=----WebKitFormBoundarygsa4MvTkWwmDzAzn
Cookie: django_language=en; csrftoken=RCTDoTVRLlWU75s902mOYSXOdhSSw7PB; sessionid=rypo0vq8c0w292x481l9ngixeb3rbkzm
DNT: 1
Host: docker.domain.ad:8000
Origin: http://docker.domain.ad:8000
Referer: http://docker.domain.ad:8000/extras/scripts/77/
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36 Edg/127.0.0.0

Response:
HTTP/1.1 302 Found
Content-Type: text/html; charset=utf-8
Location: /extras/scripts/results/161/
X-Request-ID: d52dd1a8-e7a0-4f77-b3c6-dc0bfbafbf51
Vary: HX-Request, Accept-Language, Cookie, origin
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin
Cross-Origin-Opener-Policy: same-origin
X-Frame-Options: SAMEORIGIN
Content-Length: 0
Content-Language: en
access-control-allow-origin: *
Server: Unit/1.32.0
Date: Wed, 31 Jul 2024 09:06:28 GMT
```

**Reply by Cennepal** [reply]
Alright, I set dns_name to "dummyname" and it still got reverted. Also, another discovery. When clicking "Run Again" from the scripts page, it reverts every time.

**Reply by candlerb** [reply]
You need to look at the POST *body* in those two cases, to see if different settings are being passed for the commit changes flag (or maybe it's present in one and absent in the other)

**Reply by Cennepal** [reply]
Ah, sorry. Just checked and yes it's different. Also it seems like my earlier fail with the "dummyname" was due to using "Run Again", which seems to fail 100% of the time, even with commit_default set to True.  So it probably is something with the DNS reverse lookup.
## So here is the POST body for a successful scan:
```
csrfmiddlewaretoken: LpSSzghTYYCOPT43t7sVbnCSSsOGNvuIWD41PCndeIDrHDWt5Juh6ORlrKPtl1FS
_schedule_at: 
_interval: 
_commit: 
_commit: on
_run: 
```
## POST body of a randomly failed scan:
```
csrfmiddlewaretoken: VIFlXjSVdsUilKKqRvKqBYp74IFIN5BRGCw7ae2JWCQfkyS0HBT6hGvKVTaCXnhC
_schedule_at: 
_interval: 
_commit: 
_run: 
```
## POST body using "Run Again" button:
```
csrfmiddlewaretoken: pvuIhhlFPaoM7mHhfFZaDLcYEmmmLaFsapluucvtykkJ6aPR5L8QjtiBvxRgVsld
_run: 
```

**Reply by candlerb** [reply]
```
_commit: 
_commit: on
```

Interesting. HTML shows (on my system):

```
      <div class="form-check mb-0">
        <input type="hidden" name="_commit" value="">
<input type="checkbox" name="_commit"
 aria-describedby="id__commit_helptext" id="id__commit" checked
 class="form-check-input">
        <label for="id__commit" class="form-check-label">
          Commit changes
        </label>

          <span class="form-text">Commit changes to the database (uncheck for a dry-run)</span>

      </div>
```

That is: there's a hidden field and a checkbox, both with name `_commit`.  That's pretty ugly.  Having said that, when the box has been pre-selected as 'checked' I'd expect it to return 'on', and toggling it twice in the UI shouldn't change this; but I don't know enough about the semantics of HTML forms to say for sure.  All this doesn't explain why you have this problem with one script but not others.

FWIW, the code in Netbox uses the *last* value of `_commit` supplied by the client:

```
netbox/extras/views.py:                commit=form.cleaned_data.pop('_commit')
```

Therefore, having both the hidden field and the form field is arguably not an outright bug, but it depends on the browser sending the fields in a deterministic order, and this seems unnecessary IMO.

I haven't traced to see what it does when the value is blank, but I think it should be logged:

```
    logger.info(f"Running script (commit={commit})")
```

If the raw string value is passed around, then empty string would be treated as if it were False:

```
                    if not commit:
                        raise AbortTransaction()
```

**Reply by jeremystretch** [reply]
> That is: there's a hidden field and a checkbox, both with name `_commit`. That's pretty ugly.

Ugly, but necessary in certain situations to force the browser to convey false values for a checkbox. Without the hidden field, the value is omitted from the request (rather than being included as "false"). This breaks e.g. dynamic form rendering; see bug #13064.

**Reply by Cennepal** [reply]
Yeah, it's weird. Not to mention that the "Run Again" button doesn't return `_commit` at all. I'm going to rewrite the lookup to use dnspython (just in case) and put in an if statement that catches if commit isn't true and aborts. Other than that I haven't a clue.

**Reply by Cennepal** [reply]
I have switched to using dnspython for resolution and put in a condition to catch the commit not being set. Everything works *fine* for now, but I'm still a bit paranoid. The "Run Again" button is not so important, but it's still somewhat troubling that it won't pass commit at all.  

@candlerb Thank you very much for your time and help regarding this. You saved me a lot of time :D
